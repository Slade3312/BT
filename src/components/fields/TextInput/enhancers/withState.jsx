import React from 'react';
import PropTypes from 'prop-types';
import CustomPropTypes from 'utils/prop-types';
import withForwardedRef from 'enhancers/withForwardedRef';


export default function withState(WrappedInput) {
  class StatefulInput extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        /** if value is missing, empty string is used by default */
        value: 'value' in props ? props.value : '',
      };
    }

    handleChange = (value, event) => {
      const { onChange } = this.props;

      this.setState({ value });
      if (onChange) onChange(value, event);
    };

    render() {
      const { value } = this.state;
      const { forwardedRef, ...otherProps } = this.props;
      return (
        <WrappedInput
          {...otherProps}
          ref={forwardedRef}
          onChange={this.handleChange}
          value={value}
        />
      );
    }
  }

  StatefulInput.propTypes = {
    /**
     * onChange is used to maintain value state
     * while `value` acts like initial value
     */
    value: PropTypes.any,
    onChange: PropTypes.func,

    /** is not an actual prop */
    forwardedRef: CustomPropTypes.ref,
  };


  /**
   * Override component name by prepending `Stateful~`
   * to make it look nice, for example: `StatefulTextInput`
   */
  if (process.env.NODE_ENV !== 'production') {
    const WrappedComponentName = WrappedInput.displayName || WrappedInput.name || 'Input';
    StatefulInput.displayName = `Stateful${WrappedComponentName}`;
  }

  return withForwardedRef(StatefulInput);
}
