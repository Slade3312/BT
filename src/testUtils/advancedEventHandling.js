const KEYBOARD_BACKSPACE_KEYCODE = 8;


/**
 * Inject the following into enzyme node:
 * - keypress, and keyboard events that change input value,
 *   manipulate selection and call keydown, keyup, change
 * - selectionStart, selectionEnd and setSelectionRange
 *
 * Usage:
 * ----
 * const component = mount(...);
 * const input = component.find('input');
 * injectEventHandling(input);
 * input.simulate('keypress', { key: 'a' });
 * input.simulate('keypress', { key: 'Backspace' });
 * input.simulate('keyboard', { keys: ['b', 'c', 'Backspace', 'e' ] });
 *
 * Note:
 * ----
 * This module does not exactly reproduce the default browser behavior,
 * it is simply a close approximation with limited functionality
 *
 * Amongst limitations - async nature of cursor position management
 */
export const injectEventHandling = (testNode) => {
  const originalSimulate = testNode.simulate;

  const domNode = testNode.instance();
  Object.defineProperty(domNode, 'selectionStart', {
    writable: true,
    value: 0,
  });
  Object.defineProperty(domNode, 'selectionEnd', {
    writable: true,
    value: 0,
  });

  Object.defineProperty(domNode, 'setSelectionRange', {
    writable: true,
    value: (selectionStart, selectionEnd) => {
      console.assert(
        selectionStart >= 0 && selectionEnd >= selectionStart && selectionEnd <= domNode.value.length,
        'Attempting to call setSelectionRange with improper interval',
      );

      domNode.selectionStart = selectionStart;
      domNode.selectionEnd = selectionEnd;
    },
  });

  testNode.simulate = (eventType, event, ...args) => {
    if (eventType === 'keyboard') {
      /** shortcut event type for user input simulation */
      const { keys } = event;

      keys.forEach(key => testNode.simulate('keypress', { key }));
    } else if (eventType === 'keypress') {
      /** smart keypress, supporting different keys as well as cursor movement */
      let preventDefault;
      event.preventDefault = () => { preventDefault = true; };
      if (event.key === 'Backspace') event.keyCode = KEYBOARD_BACKSPACE_KEYCODE;

      originalSimulate.call(testNode, 'keydown', event, ...args);
      /** must read instance after keydown, otherwise we miss out on changes done by keydown */
      let { value } = testNode.instance();

      switch (event.key) {
        case 'Backspace':
          /**
           * Handles three possible cases for backspace:
           * 1. if selection is empty and we are at start - we do nothing
           * 2. if selection is empty and we are not at start - we erase one symbol backwards
           * 3. if selection is not empty, we erase selection
           */
          if (domNode.selectionStart === domNode.selectionEnd && domNode.selectionStart > 0) {
            domNode.selectionStart -= 1;
          }

          value = value.slice(0, domNode.selectionStart) + value.slice(domNode.selectionEnd);
          domNode.value = value;
          domNode.setSelectionRange(domNode.selectionStart, domNode.selectionStart);

          if (!preventDefault) {
            testNode.simulate('change', { target: { value } });
          }
          break;
        default:
          /** otherwise we just enter a key */
          value = value.slice(0, domNode.selectionStart) + event.key + value.slice(domNode.selectionEnd);
          domNode.value = value;
          domNode.setSelectionRange(domNode.selectionStart + 1, domNode.selectionStart + 1);


          if (!preventDefault) {
            originalSimulate.call(testNode, 'change', { target: { value } });
          }
          break;
      }
      originalSimulate.call(testNode, 'keypress', event, ...args);
      originalSimulate.call(testNode, 'keyup', event, ...args);

      /**
       * this is default react behavior: if expected value doesn't match actual value,
       * aka onChange modifies field - we reset cursor position to last symbol
       */
      // if (value !== domNode.value) {
      //   console.log('BREAK ALL TO ASHES')
      //   domNode.setSelectionRange(domNode.value.length, domNode.value.length);
      // }
    } else if (eventType === 'paste') {
      let { value } = testNode.instance();

      value = value.slice(0, domNode.selectionStart) + event._value + value.slice(domNode.selectionEnd);
      domNode.value = value;
      const nextCursorPos = domNode.selectionStart + event._value.length;
      domNode.setSelectionRange(nextCursorPos, nextCursorPos);

      originalSimulate.call(testNode, 'change', { target: { value } });
    }

    /** keyboard is our abstract event, no propagation needed */
    if (eventType !== 'keyboard' && eventType !== 'paste') {
      originalSimulate.call(testNode, eventType, event, ...args);
    }
  };
};
