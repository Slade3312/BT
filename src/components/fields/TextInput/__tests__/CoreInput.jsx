import React from 'react';
import { mount } from 'enzyme';
import { injectEventHandling } from 'testUtils/advancedEventHandling';
import { CoreInput } from '../index';
import withState from '../enhancers/withState';

const StatefulCoreInput = withState(CoreInput);


const setupTest = (initialValue, enteredValue = '', trimInitialValue = false) => {
  const testFn = jest.fn();
  const component = mount(<StatefulCoreInput
    onChange={testFn}
    value={initialValue || ''}
    trimInitialValue={trimInitialValue}
    isClearable
  />);

  const input = component.find('input');
  injectEventHandling(input);
  const domNode = input.instance();
  domNode.setSelectionRange(initialValue.length, initialValue.length);
  input.simulate('keyboard', { keys: enteredValue.split('') });
  const icon = component.find('button');

  return { input, domNode, icon, testFn };
};


describe('CoreInput', () => {
  // console.log = () => {};
  // console.assert = () => {};
  it('Should enter value properly', () => {
    const { domNode } = setupTest('qwerty123');
    expect(domNode.value).toEqual('qwerty123');
  });

  describe('Enter and clear value', () => {
    const shouldCleanValue = (initialValue, enteredValue) => {
      initialValue = initialValue || '';
      enteredValue = enteredValue || '';
      it(`Should clean value '${initialValue}' after entering '${enteredValue}'`, () => {
        const { icon, domNode, testFn } = setupTest(initialValue, enteredValue);
        expect(domNode.value).toEqual(initialValue + enteredValue);
        expect(testFn.mock.calls.map(([a]) => a)).toMatchSnapshot();
        icon.simulate('click');
        expect(domNode.value).toEqual('');
      });
    };

    /** test initial value */
    shouldCleanValue('1');
    shouldCleanValue('q ');
    shouldCleanValue('qw ');

    /** test changed value */
    shouldCleanValue(null, 'q');
    shouldCleanValue(null, 'q ');
    shouldCleanValue(null, 'qw ');

    /** test mixed value */
    shouldCleanValue('q', ' w');
    shouldCleanValue(' sq', ' we ');
    shouldCleanValue('qw', ' e ');
    shouldCleanValue(' qw', ' e ');
  });
});
