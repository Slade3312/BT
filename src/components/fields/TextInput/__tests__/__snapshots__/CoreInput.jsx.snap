// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreInput Enter and clear value Should clean value ' qw' after entering ' e ' 1`] = `
Array [
  "qw e",
]
`;

exports[`CoreInput Enter and clear value Should clean value ' sq' after entering ' we ' 1`] = `
Array [
  "sq w",
  "sq we",
]
`;

exports[`CoreInput Enter and clear value Should clean value '' after entering 'q ' 1`] = `
Array [
  "q",
]
`;

exports[`CoreInput Enter and clear value Should clean value '' after entering 'q' 1`] = `
Array [
  "q",
]
`;

exports[`CoreInput Enter and clear value Should clean value '' after entering 'qw ' 1`] = `
Array [
  "q",
  "qw",
]
`;

exports[`CoreInput Enter and clear value Should clean value '1' after entering '' 1`] = `Array []`;

exports[`CoreInput Enter and clear value Should clean value 'q ' after entering '' 1`] = `Array []`;

exports[`CoreInput Enter and clear value Should clean value 'q' after entering ' w' 1`] = `
Array [
  "q w",
]
`;

exports[`CoreInput Enter and clear value Should clean value 'qw ' after entering '' 1`] = `Array []`;

exports[`CoreInput Enter and clear value Should clean value 'qw' after entering ' e ' 1`] = `
Array [
  "qw e",
]
`;
